source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
+    else if (linkage == spv::LinkageType::Export)
+      exports.emplace(reinterpret_cast<const char*>(insn.GetOperand(2u).words.data()), insn.GetSingleWordOperand(0u));
+    else
+      return 1;
   }
-  SetWord(merged_binary, merged_length, (3u << spv::WordCountShift) + 14);
-  SetWord(merged_binary, merged_length + 1u, static_cast<spirv_word>(addressing_model));
-  SetWord(merged_binary, merged_length + 2u, static_cast<spirv_word>(memory_model));
-  merged_length += 3u;
-
-  // Copy OpEntryPoint
-  CopySections(IsEntryPointOpcode);
-
-  // Copy OpExecutionMode
-  CopySections(IsExecutionModeOpcode);
-
-  // Copy debug a)
-  CopySections(IsDebugAOpcode);
-  // Copy debug b)
-  CopySections(IsDebugBOpcode);
-
-  // Copy annotations
-  for (uint32_t i = 0u; i < num_binaries; ++i) {
-    const uint32_t* const binary = binaries[i].data();
-    const size_t length = binaries[i].size();
-    uint32_t index = indices[i];
-    uint32_t j = index;
-    uint32_t prev_j = j;
-    while (j < length) {
-      uint32_t word_count;
-      const spv::Op opcode = ParseOpcode(binary, j, word_count);
-
-      prev_j = j;
 
-      if (opcode != spv::Op::OpDecorate && opcode != spv::Op::OpMemberDecorate &&
-          opcode != spv::Op::OpGroupDecorate && opcode != spv::Op::OpGroupMemberDecorate &&
-          opcode != spv::Op::OpDecorationGroup)
-        break;
-
-      j += word_count;
-
-      if (opcode != spv::Op::OpDecorate)
-        continue;
-      const spv::Id id = GetWord(binary, prev_j + 1u);
-      const spv::Decoration decoration = static_cast<spv::Decoration>(GetWord(binary, prev_j + 2u));
-      if (decoration == spv::Decoration::BuiltIn)
-        builtins.push_back(id);
-      if (decoration != spv::Decoration::LinkageAttributes)
-        continue;
-
-      const spv::LinkageType linkage_type = static_cast<spv::LinkageType>(GetWord(binary, prev_j + (word_count - 1u)));
-      if (linkage_type == spv::LinkageType::Export)
-        exports.emplace(id, GetString(binary, prev_j + 3u));
-      else
-        imports.emplace(id, GetString(binary, prev_j + 3u));
-
-      // Remove exports from SPIR-V if we are not building a library
-      if (create_library && linkage_type == spv::LinkageType::Export)
-        continue;
-
-      std::memcpy(merged_binary + merged_length, binary + index,
-                  (prev_j - index) * sizeof(spirv_word));
-      merged_length += prev_j - index;
-      index = j;
+  for (const auto& i : imports)
+    std::cout << "Import: " << i.first << ", " << i.second << "\n";
+  for (const auto& i : exports)
+    std::cout << "Export: " << i.second << ", " << i.first << "\n";
+
+  std::unordered_map<spv::Id, spv::Id> linking_table;
+  linking_table.reserve(imports.size());
+  for (const auto& i : imports) {
+    auto j = exports.find(i.second);
+    if (j == exports.end()) {
+      std::cerr << "Failed to find \"" << i.second << "\"\n";
```

Comment:
Diagnostic through the context object.  You'll be able to write sensible tests that way.